### PR TITLE
postpone ftpm initialization

### PR DESCRIPTION
**Pull request by muvarov**
add rfc patch to postpone ftpm initialization
until user space part will be ready.

Signed-off-by: Maxim Uvarov <maxim.uvarov@linaro.org>

### CODE REVIEW

**Comment by muvarov** [comment]
This RFC patch obsoletes https://github.com/Linaro/meta-ledge/pull/198

File: meta-ledge-bsp/recipes-kernel/linux/linux-ledge/0007-ftpm-postpone-tmp-initialization-until-userland-is-r.patch [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+From: Maxim Uvarov <maxim.uvarov@linaro.org>
+Date: Thu, 26 Mar 2020 19:17:42 +0000
+Subject: [PATCH] ftpm: postpone tmp initialization until userland is ready
+
+ftpm emulated by userland and it can fail if:
+1. tee-suplicant application is not run.
+2. ftpm TA file is missing on the file system (not yet mounted).
+This patch postpones tpm initialization with periodic checks
+for available resources.
+
```
Comment (apalos) [inline meta-ledge-bsp/recipes-kernel/linux/linux-ledge/0007-ftpm-postpone-tmp-initialization-until-userland-is-r.patch]: I don't think it's best practice to postpone module init with timers and/or scheduled workqueues in kernel. 
I don't remember the details but I think there's a different mechanism for that

File: meta-ledge-bsp/recipes-kernel/linux/linux-ledge/0007-ftpm-postpone-tmp-initialization-until-userland-is-r.patch [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+ 
+ 	/* Open context with TEE driver */
+ 	pvt_data->ctx = tee_client_open_context(NULL, ftpm_tee_match, NULL,
+ 						NULL);
+ 	if (IS_ERR(pvt_data->ctx)) {
+-		if (ERR_PTR(pvt_data->ctx) == -ENOENT)
+-			return -EPROBE_DEFER;
+ 		dev_err(dev, "%s: tee_client_open_context failed\n", __func__);
+-		return ERR_PTR(pvt_data->ctx);
++		return -EINVAL;
```
Comment (apalos) [inline meta-ledge-bsp/recipes-kernel/linux/linux-ledge/0007-ftpm-postpone-tmp-initialization-until-userland-is-r.patch]: What will happen if the supplicant never comes up? We'll keep polling?

**Comment by cpriouzeau** [comment]
My General point of view are that it's not a kernel issue but a userland problematic and a synchronization must be made on userland to load this kernel module (optee-ftpm) when tee-supplicant are loaded. It's not the responsibility of the kernel module to wait the filesystem and wait a specific service. Other point, How do you do if the service tee-supplicant doesn't start (in case of optee-os doesn't respond - not loaded on system or not optee kernel module) ?

File: meta-ledge-bsp/recipes-kernel/linux/linux-ledge/0007-ftpm-postpone-tmp-initialization-until-userland-is-r.patch (hunk 2 above)
Comment (muvarov) [reply meta-ledge-bsp/recipes-kernel/linux/linux-ledge/0007-ftpm-postpone-tmp-initialization-until-userland-is-r.patch]: yes. But it's expected to come up some time later if valid tmp entry exist in dtb.

**Comment by muvarov** [comment]
> My General point of view are that it's not a kernel issue but a userland problematic and a synchronization must be made on userland to load this kernel module (optee-ftpm) when tee-supplicant are loaded. It's not the responsibility of the kernel module to wait the filesystem and wait a specific service. 

Actually I agree with this statement. But that looks a little bit unnatural when kernel modules needed to be loaded manually after user land app starts.

>Other point, How do you do if the service tee-supplicant doesn't start (in case of optee-os doesn't respond - not loaded on system or not optee kernel module) ?

That will be polling to start forever. Actually if tee-supplicant did not start that means that you can not use things based on TPM. One way maybe to wait until tee-supplicant opens /dev/teepriv0  then start ftpm initialization.  But that looks for me over-complicated code and mix tee and ftpm drivers.

Also with this patch and without I see problem that user can kill tee-supplicant application after ftpm driver created tmp device. And I'm not sure if ftpm will operate in that case.

File: meta-ledge-bsp/recipes-kernel/linux/linux-ledge/0007-ftpm-postpone-tmp-initialization-until-userland-is-r.patch (hunk 1 above)
Comment (muvarov) [reply meta-ledge-bsp/recipes-kernel/linux/linux-ledge/0007-ftpm-postpone-tmp-initialization-until-userland-is-r.patch]: @apalos one way might be similar like this:
#define BR_STP_PROG     "/sbin/bridge-stp"
static int br_stp_call_user(struct net_bridge *br, char *arg)
```
{
..

        /* call userspace STP and report program errors */
        rc = call_usermodehelper(BR_STP_PROG, argv, envp, UMH_WAIT_PROC);
}
```
But I don't like this due to security reasons. (kernel loads some binary from fs).

**Pull request closed without merging**